### PR TITLE
Convert KubeAPI Load Balancer to the ops framework

### PR DESCRIPTION
**Pull request by mateoflorido**
# Summary
This PR showcases the migration of the KubeAPI Load Balancer reactive charm to the operator framework.

## Changes
- Added NGINX Library: Included the NGINX library within the charm to facilitate control over the lifecycle of the NGINX application, i.e. installation and configuration.
- PoC HACluster: Implemented a HACluster library as a proof of concept, which may be beneficial in future use cases.
- COS Instrumentation: Replaced the Nagios checks with alerts using the Prometheus Alert Manager from the COS lite stack. There is an open [issue](https://github.com/canonical/grafana-agent-snap/issues/27) regarding the `node-exporter` not exporting the required metrics for the `systemd` integration, but once it's fixed, the alerts should work as expected.
- Utilisation of Reconcile and Contextual Status: This charm now utilises both the [`charm-lib-reconciler`](https://github.com/charmed-kubernetes/charm-lib-reconciler/tree/main) and [`charm-lib-contextual-status`](https://github.com/charmed-kubernetes/charm-lib-contextual-status/tree/main) libraries.
- Updated NGINX Config Template: The nginx configuration template has been updated with the recommended values from [Mozilla](https://ssl-config.mozilla.org/#server=nginx&version=1.17.7&config=intermediate&openssl=3.0.2&ocsp=false&guideline=5.7). Additionally, some deprecated options (e.g., `ssl on`) have been removed.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,9 @@
...
+*.charm
 .tox/
+.coverage
 __pycache__/
-*.pyc
-*.charm
\ No newline at end of file
+*.py[cod]
+.idea
+.vscode/
```
Comment (addyess) [inline .gitignore:9]: I err on the style to not put editors in a .gitignore -- but to add thees to the global gitignore

What do you think?

```suggestion
```

File: src/charm.py [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+            dns_records = self.config.get("ha-cluster-dns").split()
+            if dns_records:
+                return dns_records
+        return []
+
+    def _get_public_address(self):
+        """Return the unit public-address."""
+        process = subprocess.run(
+            ["unit-get", "public-address"], capture_output=True, text=True, timeout=25, check=True
+        )
```
Comment (addyess) [inline src/charm.py:179]:
```suggestion
        process = subprocess.check_output(
            ["unit-get", "public-address"], text=True, timeout=25
        )
```

File: src/charm.py [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+        if forced_lb_ips:
+            return forced_lb_ips
+        if self.hacluster.is_ready:
+            vips = self.config.get("ha-cluster-vip").split()
+            if vips:
+                return vips
+            dns_records = self.config.get("ha-cluster-dns").split()
+            if dns_records:
+                return dns_records
+        return []
```
Comment (addyess) [inline src/charm.py:173]: Hooray walrus (`:=`)

```suggestion
        if forced_lb_ips := self.config.get("loadbalancer-ips").split():
            return forced_lb_ips
        if self.hacluster.is_ready:
            if vips := self.config.get("ha-cluster-vip").split():
                return vips
            if dns_records := self.config.get("ha-cluster-dns").split():
                return dns_records
        return []
```

**Review by addyess — CHANGES_REQUESTED** [review]
This is a nice port.  I've got a few things i might like some clarification on but i understand where you're going here.  Thanks @mateoflorido

File: .gitignore (hunk 1 above)
Comment (mateoflorido) [reply .gitignore:9]: Yeah, I agree. I think `charmcraft init` includes these, so I'll remove them.

File: tests/data/charm.yaml [hunk 4]
```diff
@@ -1,10 +1,8 @@
...
+    options:
+      allow-privileged: "true"
   kubeapi-load-balancer:
-    charm: {{charm}}
-    num_units: 1
-    expose: true
-relations:
-- [kubeapi-load-balancer:lb-consumers, kubernetes-control-plane:loadbalancer-internal]
-- [kubeapi-load-balancer:lb-consumers, kubernetes-control-plane:loadbalancer-external]
-- [kubeapi-load-balancer:certificates, easyrsa:client]
```
[the remark this answers is not in the document]
Comment (mateoflorido) [reply tests/data/charm.yaml:10]: I'm using the CK bundle and replacing the KubeAPI Load Balancer charm with the one it builts locally. After giving it some thought, I believe I should rollback the current integration tests, as deploying with the kuberenetes core bundle seems lighter to me.

File: tox.ini [hunk 5]
```diff
@@ -1,59 +1,78 @@
...
-    black --check {toxinidir}/reactive {toxinidir}/tests
-    mypy --config-file={toxinidir}/tox.ini {toxinidir}/reactive
+    ruff
+    codespell
+commands =
+    # if this charm owns a lib, uncomment "lib_path" variable
+    # and uncomment the following line
+    # codespell {[vars]lib_path}
+    codespell {tox_root}
+    ruff {[vars]all_path}
```
[the remark this answers is not in the document]
Comment (mateoflorido) [reply tox.ini:46]: Yes, charmcraft uses [ruff](https://github.com/astral-sh/ruff) by default.
Hmm, I can check out `mypy`  and see if I can integrate it.

File: tests/integration/test_charm.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    await ops_test.model.wait_for_idle(status="active", timeout=10 * 60)
+
+    try:
+        worker = ops_test.model.applications["kubernetes-worker"]
+        action = await worker.units[0].run("cat /root/cdk/kubeproxyconfig | grep server")
+        result = await action.wait()
+        assert f"https://{address}" in result.results["stdout"]
+    finally:
+        await api_lb.reset_config(["loadbalancer-ips"])
+        await ops_test.model.wait_for_idle(status="active", timeout=10 * 60)
```
[the remark this answers is not in the document]
Comment (mateoflorido) [reply tests/integration/test_charm.py:61]: It is my understanding that these tests used a deprecated relation to communicate with the KCP|KW. I'm unsure if we should keep them.

File: metadata.yaml [hunk 7]
```diff
@@ -19,8 +16,12 @@ series:
 requires:
   apiserver:
     interface: http
+  certificates:
+    interface: tls-certificates
```
[the remark this answers is not in the document]
Comment (mateoflorido) [reply metadata.yaml:20]: Yes, the charm uses the legacy relation.

File: charmcraft.yaml [hunk 8]
```diff
@@ -0,0 +1,16 @@
+# This file configures Charmcraft.
+# See https://juju.is/docs/sdk/charmcraft-config for guidance.
+
+type: charm
+bases:
+  - build-on:
+    - name: ubuntu
+      channel: "22.04"
```
[the remark this answers is not in the document]
Comment (mateoflorido) [reply charmcraft.yaml:8]: I don't think so. Which architectures should we target? I think we should support `amd64` and `arm64` architectures.

[the remark this answers is not in the document]
Comment (addyess) [reply charmcraft.yaml:8]: Alrighty if there's NO binary python packages... then you probably need this:

builds on amd64, runs on the three arch we support with charmed kubernetes

```yaml
bases:
  - build-on:
    - name: ubuntu
      channel: "20.04"
      architecture: ["amd64"]
    run-on:
    - name: ubuntu
      channel: "20.04"
      architecture: ["amd64", "arm64", "s390x']
    - name: ubuntu
      channel: "22.04"
      architecture: ["amd64", "arm64", "s390x']
```

File: templates/apilb.conf [hunk 9]
```diff
@@ -10,15 +10,16 @@ server {
...
-    ssl_session_cache builtin:1000 shared:SSL:10m;
+    ssl_session_cache shared:SSL:10m;
     ssl_certificate {{ server_certificate }};
     ssl_certificate_key {{ server_key }};
-    ssl_ciphers HIGH:!aNULL:!eNULL:!EXPORT:!CAMELLIA:!DES:!MD5:!PSK:!RC4;
-    ssl_prefer_server_ciphers on;
+    ssl_protocols               TLSv1.2 TLSv1.3;
+    ssl_ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:DHE-RSA-CHACHA20-POLY1305;
+    ssl_prefer_server_ciphers   on;
+    ssl_session_tickets off;
```
[the remark this answers is not in the document]
Comment (mateoflorido) [reply templates/apilb.conf:22]: Sure, Mozilla has various configurations for server software like NGINX, Apache, and others. These configurations are tailored to be compatible with different servers and clients that integrate with them. I noticed some deprecation notices in the NGINX logs, so I decided to update the template using the [intermediate](https://ssl-config.mozilla.org/#server=nginx&version=1.17.7&config=intermediate&openssl=3.0.2&ocsp=false&guideline=5.7) configuration recommended by Mozilla.

File: src/charm.py [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+        servers = {}
+        for request in self.load_balancer.all_requests:
+            for server_port in request.port_mapping.keys():
+                service: set = servers.setdefault(server_port, set())
+                service.update(
+                    (backend, backend_port)
+                    for backend, backend_port in itertools.product(
+                        request.backends, request.port_mapping.values()
+                    )
+                )
```
[the remark this answers is not in the document]
Comment (mateoflorido) [reply src/charm.py:125]: Nice 🏌🏻, hole-in-one. Thanks!

**Review by addyess — APPROVED** [review]
Lookin nice!  Awesome work

File: tests/integration/test_charm.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Cynerva) [reply tests/integration/test_charm.py]: I don't think we should. The kube-api-endpoint relation has been deprecated for years.

[the remark this answers is not in the document]
Comment (mateoflorido) [reply tests/integration/test_charm.py]: Got it, thanks @Cynerva

File: lib/charms/observability_libs/v0/juju_topology.py [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+    def __init__(self, uuid: str):
+        self.message = "'{}' is not a valid UUID.".format(uuid)
+        super().__init__(self.message)
+
+
+class JujuTopology:
+    """JujuTopology is used for storing, generating and formatting juju topology information.
+
+    DEPRECATED: This class is deprecated. Use `pip install cosl` and
+    `from cosl.juju_topology import JujuTopology` instead.
```
Comment (Cynerva) [inline lib/charms/observability_libs/v0/juju_topology.py:93]: Can we follow this deprecation warning and remove this file? I don't see it being imported anywhere. It looks like the cos_agent lib already has `cosl` in `PYDEPS` and is getting JujuTopology from there.

File: config.yaml
Comment (Cynerva) [inline config.yaml:1]: It looks like the `port` config option is totally unused now, and could be removed from config.yaml.

**Review by Cynerva — CHANGES_REQUESTED** [review]
Awesome work. This does look to be a fully functioning replacement to the reactive version of this charm.

I do want to see the unused `port` config option removed, just to clean up the config documentation. I noticed a couple other things that could use some cleanup (see other comments about `site.toml` and `JujuTopology`) but would not consider those blockers.

**Comment by mateoflorido** [comment]
> Awesome work. This does look to be a fully functioning replacement to the reactive version of this charm.
> 
> I do want to see the unused `port` config option removed, just to clean up the config documentation. I noticed a couple other things that could use some cleanup (see other comments about `site.toml` and `JujuTopology`) but would not consider those blockers.

@Cynerva. I've removed the `port` config option and tidied up the parts you pointed out. Thank you for the review, the code looks cleaner now 😃 .

**Pull request merged**